### DISCUSSION TITLE
Add MODE2 (as MCP23008) on DIGTIAN Boards Input IO.

### DISCUSSION THREAD

**Discussion by miky2k** [arendst/Tasmota#21932] [Ideas]
HI,
on DIGTIAN relay board is used a sort of I/O extender that can be configured in the tasmota driver at compile time as INx,SWITCHx or BUTTONx,  all the ports are configured with same functionality.
It would be interesting to add a MODE2 where a file can describe every InputIO port as INx SWITCHx and BUTTONx separately.
I'm not a  skilled programmer (i try only to  define at compile time different behaviors )  but i suppose a skilled programmer can copy MCP code and adapt to DIGTIAN I/O very fast.

**Comment by Jason2866** [comment]
Not possible. There is no port expander used. It is only a shift register.

**Reply by miky2k** [reply]
Every 50ms i read 8 bit from shift register to memory, after that i can choice to use all 8 bit as button ,input or switch, i cannot find a reason why is not possible to use different bits for different  devices (buttons/switch or input), at compile time is possible to change all 8 bit behavior.

**Reply by barbudor** [reply]
The hardware is fixed : there is 8 inputs AND 8 outputs (or 16+16 or 32+32 depending on the model of board)
It is not a controller like the MCP where you can choose for each GPIO if it is an input or an output.
So you have:
- 8 relay outputs
and
- 8 inputs

Inputs can be either switch or button but the choice has been made to design the driver to use either all inputs as switches or all inputs as buttons.
Considering that a switch cover 90% of a button feature, I don't see any interrest in choosing bit-per-bit switch or button.

What is your use case?

**Reply by miky2k** [reply]
I need some button  and some binary  sensors  (INPUT in DINGTIAN  driver syntax) ,  i could not find a way to query input status if is assigned to button or a switch.

**Reply by barbudor** [reply]
You cannot query an input on tasmota. Tasmota reports the state and/or the change of state.
You should use "switch" configuration and use the proper `Switchmode`
use `Switchmode<x> 15`  for the binary sensors 
and `Switchmode<x> 4` to get a behavior similar to a button

**Reply by miky2k** [reply]
I need this  (taken from my test board)
22:08:05.962 CMD: status 10
22:08:05.971 MQT: stat/testtappa_532E44/STATUS10 = {"StatusSNS":{"Time":"2024-08-08T22:08:05","DINGTIAN":{"IN1":0,"IN2":0,"IN3":0,"IN4":0,"IN5":0,"IN6":0,"IN7":0,"IN8":0}}}

But only for defined inputs ( i use teleperiod to read input status)  , the other could be switch or buttons

**Reply by Jason2866** [reply]
Via rules the inputs can be handled you want to react on

**Reply by miky2k** [reply]
> Via rules the inputs can be handled you want to react on

Is more simple add a define that leave INPUT code enabled when switch or buttons are selected.
A change in input is routed through mqtt as sensor  and button or switch logic is working for other purposes (more simple when there is shutter enabled).

I would try this way

**Reply by miky2k** [reply]
```

/********************************************************************************************************
 * Driver operations
 */

void DingtianLoop()
{
  uint32_t inputs = DingtianReadWrite(Dingtian->outputs);
  uint32_t last_inputs = Dingtian->last_inputs;
  Dingtian->last_inputs = inputs;

#if !defined(DINGTIAN_USE_AS_BUTTON) && !defined(DINGTIAN_USE_AS_SWITCH) || defined(DINGTIAN_FORCE_INPUT_ENABLED)
  if (inputs != last_inputs) {
    AddLog(LOG_LEVEL_DEBUG_MORE, PSTR("DNGT: inputs=0x%08X, last=0x%08X"), inputs, last_inputs);
    bool first_done = false;
    ResponseTime_P(PSTR(",\"DINGTIAN_CHG\":{"));
    for (int i = 0 ; i < Dingtian->count ; i++, last_inputs>>=1, inputs>>=1) {
      if ((last_inputs & 1) != (inputs & 1)) {
        if (first_done) ResponseAppend_P(PSTR(","));
        ResponseAppend_P(PSTR("\"IN%d\":%d"), i +1, (inputs & 1));
        //AddLog(LOG_LEVEL_DEBUG_MORE, PSTR("\"IN%d\":%d"), i +1, (inputs & 1));
        first_done = true;
      }
    }
    ResponseAppend_P(PSTR("}}"));
    if (first_done) {
      MqttPublishPrefixTopicRulesProcess_P(STAT, PSTR("DINGTIAN_CHG"));
    }
  }
#endif   

#if defined(DINGTIAN_USE_AS_BUTTON) || defined(DINGTIAN_USE_AS_SWITCH)
  inputs = Dingtian->last_inputs;
  for (int i = 0 ; i < Dingtian->count ; i++, inputs>>=1) {
  #ifdef DINGTIAN_USE_AS_BUTTON
      ButtonSetVirtualPinState(Dingtian->key_offset +i, inputs &1);
  #else
      SwitchSetVirtualPinState(Dingtian->key_offset +i, inputs &1);
  #endif
    }
#endif                      
 
     
}
```

I modify loop and other stuff , i leave it here for documentation , perhaps in the future i would  try to pull a request.
